pkg/adaptor/mongodb/reader.go

```diff
@@ -0,0 +1,160 @@
+package mongodb
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/compose/transporter/pkg/client"
+	"github.com/compose/transporter/pkg/log"
+	"github.com/compose/transporter/pkg/message"
+	"github.com/compose/transporter/pkg/message/data"
+	"github.com/compose/transporter/pkg/message/ops"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+var (
+	_ client.Reader = &Reader{}
+)
+
+// Reader implements the behavior defined by client.Reader for interfacing with MongoDB.
+type Reader struct {
+	db string
+}
+
+func newReader(db string) client.Reader {
+	return &Reader{db}
+}
+
+type resultDoc struct {
+	doc bson.M
+	c   string
+}
+
+func (r *Reader) Read(filterFn client.NsFilterFunc) client.MessageChanFunc {
+	return func(s client.Session, done chan struct{}) (chan message.Msg, error) {
+		out := make(chan message.Msg)
+		session := s.(*Session)
+		go func() {
+			defer close(out)
+			defer log.With("db", r.db).Infoln("starting Read func")
+			cSession := session.mgoSession.Copy()
+			defer cSession.Close()
+			collections := r.listCollections(cSession, filterFn)
+			iSession := session.mgoSession.Copy()
+			defer iSession.Close()
+			results := r.iterateCollection(iSession, collections)
+			for {
+				select {
+				case <-done:
+					return
```

Comment:
This `done` doesn't propagate to the `listCollecitons` and `iterateCollection` goroutines. would be nice to clean those up as well. Although I think the application is shutting down in this instance, right? Not too big an issue.